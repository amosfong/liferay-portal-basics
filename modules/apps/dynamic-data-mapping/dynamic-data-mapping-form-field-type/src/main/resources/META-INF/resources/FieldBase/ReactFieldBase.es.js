/**
 * SPDX-FileCopyrightText: (c) 2000 Liferay, Inc. https://liferay.com
 * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
 */

import ClayButton, {ClayButtonWithIcon} from '@clayui/button';
import ClayForm from '@clayui/form';
import ClayIcon from '@clayui/icon';
import ClayLabel from '@clayui/label';
import ClayPopover from '@clayui/popover';
import classNames from 'classnames';
import {
	EVENT_TYPES as CORE_EVENT_TYPES,
	FieldFeedback,
	Layout,
	PagesVisitor,
	useForm,
	useFormState,
} from 'data-engine-js-components-web';
import {sub} from 'frontend-js-web';
import React, {useCallback, useEffect, useMemo, useState} from 'react';

import './FieldBase.scss';

export function updateFieldNameLocale(editingLanguageId, locale, name) {
	return name.replace(new RegExp(`${editingLanguageId}$`), locale);
}

export function normalizeInputValue(fieldType, value) {
	if (!value) {
		return '';
	}

	if (
		fieldType === 'document_library' ||
		fieldType === 'geolocation' ||
		fieldType === 'grid' ||
		fieldType === 'image' ||
		fieldType === 'select'
	) {
		return !Object.keys(value).length ? '' : JSON.stringify(value);
	}

	return value;
}

const getFieldDetails = ({
	errorMessage,
	hasError,
	label,
	required,
	text,
	tip,
	warningMessage,
}) => {
	const fieldDetails = [];

	if (label) {
		fieldDetails.push(Liferay.Util.escape(label));
	}

	if (tip) {
		fieldDetails.push(Liferay.Util.escape(tip));
	}

	if (text) {
		fieldDetails.push(Liferay.Util.escape(text));
	}

	if (hasError) {
		fieldDetails.push(Liferay.Util.escape(errorMessage));
	}
	else {
		if (warningMessage) {
			fieldDetails.push(Liferay.Util.escape(warningMessage));
		}
		if (required) {
			fieldDetails.push(Liferay.Language.get('required'));
		}
	}

	return fieldDetails.length ? fieldDetails.join('<br>') : false;
};

const HideFieldProperty = () => {
	return (
		<ClayLabel className="ml-1" displayType="secondary">
			{Liferay.Language.get('hidden')}
		</ClayLabel>
	);
};

const LabelProperty = ({hideField, label}) => {
	return hideField ? <span className="text-secondary">{label}</span> : label;
};

const RequiredProperty = () => {
	return (
		<span className="ddm-label-required reference-mark">
			<ClayIcon symbol="asterisk" />
		</span>
	);
};

const FieldInformation = ({popover, tooltip}) => {
	return popover ? (
		<Popover {...popover} />
	) : (
		<span
			className="c-ml-2 text-4 text-secondary"
			data-testid="tooltip"
			tabIndex={0}
			title={tooltip}
		>
			<ClayIcon symbol="question-circle-full" />
		</span>
	);
};

const Popover = ({alignPosition, content, header, image}) => {
	const [isPopoverVisible, setIsPopoverVisible] = useState(false);

	const POPOVER_MAX_WIDTH = 256;

	return (
		<ClayPopover
			alignPosition={alignPosition}
			closeOnClickOutside
			data-testid="clayPopover"
			disableScroll
			header={header}
			onShowChange={setIsPopoverVisible}
			show={isPopoverVisible}
			style={{maxWidth: POPOVER_MAX_WIDTH}}
			trigger={
				<ClayButtonWithIcon
					aria-label={Liferay.Language.get('more-information')}
					className="c-ml-2 text-secondary"
					data-testid="tooltip"
					displayType="unstyled"
					monospaced={false}
					size="sm"
					symbol="question-circle-full"
				/>
			}
		>
			<p
				className="mb-4"
				dangerouslySetInnerHTML={{
					__html: content,
				}}
			/>

			{image && (
				<img
					alt={image.alt}
					height={image.height}
					src={image.src}
					width={image.width}
				/>
			)}
		</ClayPopover>
	);
};

const FIELDSET_REGEX = /Fieldset\d+/g;
const FIELDSET_REPEAT_INDEX_REGEX = /\$(\d+)(?:#|\$|$)/g;

export default function FieldBase({
	accessible = true,
	children,
	displayErrors,
	editOnlyInDefaultLanguage,
	errorMessage,
	fieldName,
	fieldReference,
	hideField,
	hideEditedFlag,
	id,
	instanceId,
	itemPath,
	label,
	localizedValue = {},
	localizedValueEdited,
	name,
	nestedFields,
	onClick,
	overMaximumRepetitionsLimit,
	popover,
	readOnly,
	repeatable,
	required,
	showLabel = true,
	style,
	text,
	tip,
	tooltip,
	type,
	valid,
	visible,
	warningMessage,
}) {
	const {editingLanguageId, pages} = useFormState();
	const [disabledRepeatableButton, setDisabledRepeatableButton] =
		useState(false);
	const dispatch = useForm();

	const hasError = displayErrors && errorMessage && !valid;

	const fieldDetails = getFieldDetails({
		errorMessage,
		hasError,
		label,
		required,
		text,
		tip,
		warningMessage,
	});

	const fieldDetailsId = `${id ?? name}_fieldDetails`;
	const fieldLabelId = `${id ?? name}_fieldLabel`;

	const hiddenTranslations = useMemo(() => {
		if (!localizedValue) {
			return;
		}

		return Object.entries(localizedValue).map(([locale, value]) => {
			return (
				<input
					data-field-name={`${fieldName}${instanceId}`}
					data-languageid={locale}
					data-translated={
						!!localizedValueEdited?.[editingLanguageId]
					}
					key={locale}
					type="hidden"
					value={normalizeInputValue(type, value)}
					{...(locale !== editingLanguageId && {
						name: updateFieldNameLocale(
							editingLanguageId,
							locale,
							name
						),
					})}
				/>
			);
		});
	}, [
		localizedValue,
		localizedValueEdited,
		editingLanguageId,
		fieldName,
		instanceId,
		name,
		type,
	]);

	const renderLabel =
		(label && showLabel) || hideField || repeatable || required || tooltip;
	const showDisabledFieldIcon =
		editOnlyInDefaultLanguage && showLabel && readOnly;
	const showGroup =
		type === 'checkbox_multiple' ||
		type === 'grid' ||
		type === 'paragraph' ||
		type === 'radio';
	const popoverOrTooltip = !!popover || !!tooltip;
	const showFor =
		type === 'date' ||
		type === 'document_library' ||
		type === 'text' ||
		type === 'numeric' ||
		type === 'image' ||
		type === 'rich_text' ||
		type === 'search_location';
	const readFieldDetails = !showFor;
	const hasFieldDetails =
		accessible && fieldDetails && readFieldDetails && type !== 'select';

	const accessiblePropsGroup = {
		...(!renderLabel && {'aria-labelledby': fieldDetailsId}),
		...(type === 'fieldset' && {role: 'group'}),
	};

	const accessiblePropsFields = {
		...(hasFieldDetails && {'aria-labelledby': fieldDetailsId}),
		...(showFor && {htmlFor: id ?? name}),
		...readFieldDetails,
	};

	const defaultRows = nestedFields?.map((field) => ({
		columns: [{fields: [field], size: 12}],
	}));

	const checkRepetitions = useMemo(() => {
		if (repeatable && name) {
			const currentFieldFieldsets = name.match(FIELDSET_REGEX);
			const currentFieldsetRepeatIndexes = name.match(
				FIELDSET_REPEAT_INDEX_REGEX
			);

			if (currentFieldsetRepeatIndexes) {
				currentFieldsetRepeatIndexes.pop();
			}

			const visitor = new PagesVisitor(pages);

			const repeatableFields = [];

			visitor.visitFields((field) => {
				const fieldName = field.name ?? field.fieldName;

				const fieldFieldsets = fieldName.match(FIELDSET_REGEX);
				const fieldsetRepeatIndexes = fieldName.match(
					FIELDSET_REPEAT_INDEX_REGEX
				);

				if (fieldsetRepeatIndexes) {
					fieldsetRepeatIndexes.pop();
				}

				const isSameFieldset =
					currentFieldFieldsets &&
					fieldFieldsets &&
					currentFieldsetRepeatIndexes &&
					fieldsetRepeatIndexes &&
					currentFieldFieldsets.every(
						(fieldFieldset, index) =>
							fieldFieldset === fieldFieldsets[index]
					) &&
					currentFieldsetRepeatIndexes.every(
						(fieldFieldset, index) =>
							fieldFieldset === fieldsetRepeatIndexes[index]
					);

				if (fieldReference === field.fieldReference && isSameFieldset) {
					repeatableFields.push(field);
				}

				if (
					!currentFieldFieldsets &&
					fieldReference === field.fieldReference
				) {
					repeatableFields.push(field);
				}
			});

			return repeatableFields.length;
		}
	}, [fieldReference, name, pages, repeatable]);

	const disableRepeatableButton = () => {
		setDisabledRepeatableButton(true);

		setTimeout(() => {
			setDisabledRepeatableButton(false);
		}, 1000);
	};

	const translationFilterChange = useCallback(
		(event) => {
			const pagesVisitor = new PagesVisitor(pages);
			switch (event.option) {
				case 'translated':
					dispatch({
						payload: pagesVisitor.mapFields(
							(field) => {
								if (!field.localizable) {
									return {
										...field,
										disabled: true,
										hidden: true,
										visible: false,
									};
								}
								if (
									field.localizedValueEdited?.[
										editingLanguageId
									]
								) {
									return {
										...field,
										disabled: false,
										hidden: false,
										visible: true,
									};
								}
								else {
									return {
										...field,
										disabled: true,
										hidden: true,
										visible: false,
									};
								}
							},
							false,
							true
						),
						type: CORE_EVENT_TYPES.PAGE.UPDATE,
					});

					break;
				case 'untranslated':
					dispatch({
						payload: pagesVisitor.mapFields(
							(field) => {
								if (!field.localizable) {
									return {
										...field,
										disabled: true,
										hidden: true,
										visible: false,
									};
								}
								if (
									field.localizedValueEdited?.[
										editingLanguageId
									]
								) {
									return {
										...field,
										disabled: true,
										hidden: true,
										visible: false,
									};
								}
								else {
									return {
										...field,
										disabled: false,
										hidden: false,
										visible: true,
									};
								}
							},
							false,
							true
						),
						type: CORE_EVENT_TYPES.PAGE.UPDATE,
					});
					break;
				default:
					dispatch({
						payload: pagesVisitor.mapFields(
							(field) => {
								return {
									...field,
									disabled: false,
									hidden: false,
									visible: true,
								};
							},
							false,
							true
						),
						type: CORE_EVENT_TYPES.PAGE.UPDATE,
					});
					break;
			}
		},
		[dispatch, editingLanguageId, pages]
	);

	useEffect(() => {
		Liferay.on('disableRepeatableButton', disableRepeatableButton);

		return () => {
			Liferay.detach('disableRepeatableButton', disableRepeatableButton);
		};
	}, []);

	const markAsTranslated = useCallback(() => {
		const pagesVisitor = new PagesVisitor(pages);

		dispatch({
			payload: pagesVisitor.mapFields(
				(field) => {
					if (!field.localizedValue) {
						return;
					}

					return {
						...field,
						localizedValue: {
							...field.localizedValue,
							[editingLanguageId]: field.value,
						},
						localizedValueEdited: {
							...field.localizedValueEdited,
							[editingLanguageId]: true,
						},
					};
				},
				false,
				true
			),
			type: CORE_EVENT_TYPES.PAGE.UPDATE,
		});
	}, [dispatch, editingLanguageId, pages]);

	const resetTranslations = useCallback(
		({defaultLanguageId}) => {
			const pagesVisitor = new PagesVisitor(pages);

			dispatch({
				payload: pagesVisitor.mapFields(
					(field) => {
						const defaultValue =
							field.localizedValue[defaultLanguageId];
						if (field.localizedValue?.[editingLanguageId]) {
							delete field.localizedValue[editingLanguageId];
						}
						if (field.localizedValueEdited?.[editingLanguageId]) {
							delete field.localizedValueEdited[
								editingLanguageId
							];
						}

						return {
							...field,
							value: defaultValue,
						};
					},
					false,
					true
				),
				type: CORE_EVENT_TYPES.PAGE.UPDATE,
			});
		},
		[dispatch, editingLanguageId, pages]
	);

	useEffect(() => {
		Liferay.on('inputLocalized:resetTranslations', resetTranslations);
		Liferay.on('inputLocalized:markAsTranslated', markAsTranslated);
		Liferay.on(
			'inputLocalized:translationFilterChange',
			translationFilterChange
		);

		return () => {
			Liferay.detach(
				'inputLocalized:resetTranslations',
				resetTranslations
			);
			Liferay.detach('inputLocalized:markAsTranslated', markAsTranslated);
			Liferay.on('translationFilterChange', translationFilterChange);
		};
	}, [resetTranslations, markAsTranslated, translationFilterChange]);

	return (
		<ClayForm.Group
			{...accessiblePropsGroup}
			className={classNames({
				'has-error': hasError,
				'has-warning': warningMessage && !hasError,
				'hide': !visible,
			})}
			data-field-name={name}
			data-field-reference={fieldReference}
			onClick={onClick}
			style={style}
		>
			{repeatable && (
				<div className="lfr-ddm-form-field-repeatable-toolbar">
					{checkRepetitions > 1 && (
						<ClayButton
							aria-label={sub(
								Liferay.Language.get('remove-duplicate-field'),
								label ? label : type
							)}
							className={classNames(
								'ddm-form-field-repeatable-delete-button p-0',
								{
									'ddm-form-field-repeatable-button-disabled':
										disabledRepeatableButton,
								}
							)}
							disabled={readOnly || disabledRepeatableButton}
							onClick={() => {
								dispatch({
									payload: name,
									type: CORE_EVENT_TYPES.FIELD.REMOVED,
								});

								Liferay.fire('journal:storeState', {
									fieldName: Liferay.Language.get(
										'remove-repeatable-field'
									),
								});
							}}
							small
							title={Liferay.Language.get('remove')}
							type="button"
						>
							<ClayIcon symbol="hr" />
						</ClayButton>
					)}

					<ClayButton
						aria-label={sub(
							Liferay.Language.get('add-duplicate-field'),
							label ? label : type
						)}
						className={classNames(
							'ddm-form-field-repeatable-add-button p-0',
							{
								'ddm-form-field-repeatable-button-disabled':
									disabledRepeatableButton,
								'hide': overMaximumRepetitionsLimit,
							}
						)}
						disabled={readOnly || disabledRepeatableButton}
						onClick={() => {
							dispatch({
								payload: name,
								type: CORE_EVENT_TYPES.FIELD.REPEATED,
							});

							Liferay.fire('journal:storeState', {
								fieldName: Liferay.Language.get(
									'add-repeatable-field'
								),
							});
						}}
						small
						title={Liferay.Language.get('duplicate')}
						type="button"
					>
						<ClayIcon symbol="plus" />
					</ClayButton>
				</div>
			)}

			{renderLabel && (
				<>
					{showGroup ? (
						<div aria-labelledby={fieldLabelId} role="group">
							<label
								{...accessiblePropsFields}
								className={classNames('lfr-ddm-legend', {
									'text-muted': showDisabledFieldIcon,
								})}
								id={fieldLabelId}
							>
								{showLabel && label}

								{required && <RequiredProperty />}
							</label>

							{popoverOrTooltip && (
								<FieldInformation
									popover={popover}
									tooltip={tooltip}
								/>
							)}

							{showDisabledFieldIcon && (
								<FieldInformation
									tooltip={Liferay.Language.get(
										'this-field-cannot-be-localized'
									)}
								/>
							)}

							{children}
						</div>
					) : (
						<>
							<label
								{...accessiblePropsFields}
								className={classNames({
									'ddm-empty': !showLabel && !required,
									'ddm-label': showLabel || required,
									'ddm-repeatable': repeatable,
									'text-muted': showDisabledFieldIcon,
								})}
								{...(type === 'select' && {id: id ?? name})}
							>
								{showLabel && label && (
									<LabelProperty
										hideField={hideField}
										label={label}
									/>
								)}

								{required && <RequiredProperty />}

								{hideField && <HideFieldProperty />}
							</label>

							{showLabel && popoverOrTooltip && (
								<FieldInformation
									popover={popover}
									tooltip={tooltip}
								/>
							)}

							{showDisabledFieldIcon && (
								<FieldInformation
									tooltip={Liferay.Language.get(
										'this-field-cannot-be-localized'
									)}
								/>
							)}

							{children}

							{!showLabel && popoverOrTooltip && (
								<FieldInformation
									popover={popover}
									tooltip={tooltip}
								/>
							)}
						</>
					)}
				</>
			)}

			{!renderLabel && children}

			{hiddenTranslations}

			{!hideEditedFlag && (
				<input
					name={`${name}_edited`}
					type="hidden"
					value={localizedValue[editingLanguageId] !== undefined}
				/>
			)}

			<FieldFeedback
				aria-hidden={readFieldDetails}
				errorMessage={hasError ? errorMessage : undefined}
				helpMessage={typeof tip === 'string' ? tip : undefined}
				name={id ?? name}
				warningMessage={warningMessage}
			/>

			{hasFieldDetails && (
				<span
					className="sr-only"
					dangerouslySetInnerHTML={{
						__html: fieldDetails,
					}}
					id={fieldDetailsId}
				/>
			)}

			{defaultRows && <Layout itemPath={itemPath} rows={defaultRows} />}
		</ClayForm.Group>
	);
}
