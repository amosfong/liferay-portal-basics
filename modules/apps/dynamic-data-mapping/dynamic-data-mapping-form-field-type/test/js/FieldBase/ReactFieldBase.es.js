/**
 * SPDX-FileCopyrightText: (c) 2000 Liferay, Inc. https://liferay.com
 * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
 */

import {act, fireEvent, render} from '@testing-library/react';
import {FormProvider, PageProvider} from 'data-engine-js-components-web';
import React from 'react';

import '@testing-library/jest-dom/extend-expect';

import FieldBase, {
	normalizeInputValue,
	updateFieldNameLocale,
} from '../../../src/main/resources/META-INF/resources/FieldBase/ReactFieldBase.es';

const spritemap = 'icons.svg';

const FieldBaseWithProvider = (props) => (
	<FormProvider initialState={{pages: []}}>
		<PageProvider value={{editingLanguageId: 'en_US'}}>
			<FieldBase {...props} />
		</PageProvider>
	</FormProvider>
);

describe('ReactFieldBase', () => {

	// eslint-disable-next-line no-console
	const originalWarn = console.warn;

	beforeAll(() => {
		window.themeDisplay = {
			...window.themeDisplay,
			getPathThemeImages: () => 'http://localhost:8080',
		};

		// eslint-disable-next-line no-console
		console.warn = (...args) => {
			if (/DataProvider: Trying/.test(args[0])) {
				return;
			}
			originalWarn.call(console, ...args);
		};
	});

	afterAll(() => {

		// eslint-disable-next-line no-console
		console.warn = originalWarn;
	});

	beforeEach(() => {
		jest.useFakeTimers();
		fetch.mockResponseOnce(JSON.stringify({}));
	});

	it('renders the default markup', () => {
		const {container} = render(
			<FieldBaseWithProvider spritemap={spritemap} />
		);

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('renders the FieldBase with required', () => {
		const {container} = render(
			<FieldBaseWithProvider required spritemap={spritemap} />
		);

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('renders the FieldBase with id', () => {
		const {container} = render(
			<FieldBaseWithProvider id="Id" spritemap={spritemap} />
		);

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('renders the FieldBase with help text', () => {
		const {container} = render(
			<FieldBaseWithProvider
				spritemap={spritemap}
				tip="Type something!"
			/>
		);

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('renders the FieldBase with label', () => {
		const {container} = render(
			<FieldBaseWithProvider label="Text" spritemap={spritemap} />
		);

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('renders the FieldBase with tooltip', () => {
		const {findByTestId} = render(
			<FieldBaseWithProvider spritemap={spritemap} tooltip="Tooltip" />
		);

		act(() => {
			jest.runAllTimers();
		});

		expect(findByTestId('tooltip')).not.toBeNull();
	});

	it('does not render the label if showLabel is false', () => {
		const {container} = render(
			<FieldBaseWithProvider
				label="Text"
				showLabel={false}
				spritemap={spritemap}
			/>
		);

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('renders the FieldBase with contentRenderer', () => {
		const {container} = render(
			<FieldBaseWithProvider spritemap={spritemap}>
				<div>
					<h1>Foo bar</h1>
				</div>
			</FieldBaseWithProvider>
		);

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('renders the add button when repeatable is true', () => {
		const {container} = render(
			<FieldBaseWithProvider
				label="Text"
				repeatable={true}
				showLabel={false}
				spritemap={spritemap}
			/>
		);
		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('does not render the add button when repeatable is true and the maximum limit of repetions is reached', () => {
		const {container} = render(
			<FieldBaseWithProvider
				label="Text"
				overMaximumRepetitionsLimit={true}
				repeatable={true}
				showLabel={false}
				spritemap={spritemap}
			/>
		);
		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
	});

	it('shows the popover for Format field when hovering over the tooltip icon', async () => {
		const {findByTestId, getByRole, getByText} = render(
			<FieldBaseWithProvider
				fieldName="inputMaskFormat"
				popover={{
					alignPosition: 'right-bottom',
					content: 'Tooltip Description',
					header: 'input-mask-format',
					image: {
						alt: 'input-mask-format',
						height: 170,
						src: 'http://localhost:8080/forms/input_mask_format.png',
						width: 232,
					},
				}}
				spritemap={spritemap}
				tooltip="Tooltip Description"
			/>
		);

		const tooltipIcon = await findByTestId('tooltip');

		fireEvent.click(tooltipIcon);

		const clayPopover = await findByTestId('clayPopover');

		expect(clayPopover.style).toHaveProperty('maxWidth', '256px');

		expect(getByRole('img')).toHaveAttribute('height', '170');
		expect(getByRole('img')).toHaveAttribute(
			'src',
			'http://localhost:8080/forms/input_mask_format.png'
		);
		expect(getByRole('img')).toHaveAttribute('width', '232');

		expect(getByText('input-mask-format')).toBeInTheDocument();
		expect(getByText('Tooltip Description')).toBeInTheDocument();
	});

	it('renders the hidden inputs with data-languageid and data-field-name', () => {
		const localizedValue = {ca_ES: 'test_ca_ES', en_US: 'test_en_US'};

		render(
			<FieldBaseWithProvider
				fieldName="field_name"
				instanceId="instance_id"
				localizedValue={localizedValue}
				name="test_name"
			/>
		);

		const inputs = document.querySelectorAll('[name="test_name"]');

		inputs.forEach((input, i) => {
			expect(input).toHaveAttribute(
				'data-field-name',
				'field_nameinstance_id'
			);
			expect(input).toHaveAttribute(
				'data-languageid',
				Object.keys(localizedValue)[i]
			);
		});
	});

	it('renders the label with info icon and its corresponding styles when the field is non-localizable', () => {
		const {getByLabelText, getByTitle} = render(
			<FieldBaseWithProvider
				editOnlyInDefaultLanguage
				label="my-label"
				readOnly
			/>
		);

		expect(
			getByTitle('this-field-cannot-be-localized')
		).toBeInTheDocument();

		expect(getByLabelText('my-label')).toHaveClass('text-muted');
	});

	describe('Hide Field', () => {
		it('renders the FieldBase with hideField markup', () => {
			const {getAllByText, getByText} = render(
				<FieldBaseWithProvider
					hideField
					label="Text"
					spritemap={spritemap}
				/>
			);

			expect(getByText('hidden')).toBeInTheDocument();

			const allByText = getAllByText('Text');
			expect(allByText).toHaveLength(2);
			expect(allByText[0]).toBeInTheDocument();
			expect(allByText[1]).toBeInTheDocument();

			expect(getByText('hidden').parentNode).toHaveAttribute(
				'class',
				'label ml-1 label-secondary'
			);
			expect(allByText[0]).toHaveAttribute('class', 'text-secondary');
			expect(allByText[1]).toHaveAttribute('class', 'sr-only');
		});

		it('renders the FieldBase with hideField markup when the label is empty', () => {
			const {getByText} = render(
				<FieldBaseWithProvider
					hideField
					label=""
					spritemap={spritemap}
				/>
			);

			expect(getByText('hidden')).toBeInTheDocument();

			expect(getByText('hidden').parentNode).toHaveAttribute(
				'class',
				'label ml-1 label-secondary'
			);
		});
	});

	describe('updateFieldNameLocale function', () => {
		it('checks if the name only changes the language id at the end even when using a custom language', () => {

			// en_US -> language out-of-the-box
			// co -> language customized

			const customLanguageFieldName = 'com_liferay_fieldname$$co';
			const defaultLanguageFieldName = 'com_liferay_fieldname$$en_US';

			expect(
				updateFieldNameLocale('co', 'en_US', customLanguageFieldName)
			).toBe(defaultLanguageFieldName);

			expect(
				updateFieldNameLocale('en_US', 'co', defaultLanguageFieldName)
			).toBe(customLanguageFieldName);
		});
	});

	describe('normalizeInputValue function', () => {
		it('checks if the value is being formatted according to their fieldType', () => {

			// no value and any fieldType

			expect(normalizeInputValue('text', null)).toBe('');

			// text fieldType

			const textValue = 'this is a text';

			expect(normalizeInputValue('text', textValue)).toBe(textValue);

			// date and date_time fieldType

			const dateValue = '2024-12-25';
			const dateTimeValue = '2024-12-25 21:00';

			expect(normalizeInputValue('date', dateValue)).toBe(dateValue);

			expect(normalizeInputValue('date_time', dateTimeValue)).toBe(
				dateTimeValue
			);

			// image fieldType

			const imageValue = {
				alt: 'this is an alt text',
				classNameId: 22222,
				description: 'this is a description',
				fileEntryId: '33333',
				groupId: '10000',
				height: 900,
				title: 'my_image',
				type: 'document',
				url: '/documents/images/my_image',
				width: 900,
			};

			expect(normalizeInputValue('image', imageValue)).toBe(
				JSON.stringify(imageValue)
			);
		});
	});
});
