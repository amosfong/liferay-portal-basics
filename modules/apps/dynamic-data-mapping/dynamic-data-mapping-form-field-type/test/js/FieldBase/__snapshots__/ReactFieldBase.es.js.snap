// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactFieldBase does not render the add button when repeatable is true and the maximum limit of repetions is reached 1`] = `
<div>
  <div
    class="form-group hide"
  >
    <div
      class="lfr-ddm-form-field-repeatable-toolbar"
    >
      <button
        aria-label="add-duplicate-field"
        class="ddm-form-field-repeatable-add-button p-0 hide btn btn-sm btn-primary"
        title="duplicate"
        type="button"
      >
        <svg
          class="lexicon-icon lexicon-icon-plus"
          role="presentation"
        >
          <use
            href="#plus"
          />
        </svg>
      </button>
    </div>
    <label
      aria-labelledby="undefined_fieldDetails"
      class="ddm-empty ddm-repeatable"
    />
    <input
      name="undefined_edited"
      type="hidden"
      value="false"
    />
    <div
      aria-hidden="true"
      class="form-feedback-group lfr-de__field-feedback"
    >
      <div
        aria-live="assertive"
        class="form-feedback-item"
        id="undefined_fieldError"
      />
    </div>
    <span
      class="sr-only"
      id="undefined_fieldDetails"
    >
      Text
    </span>
  </div>
</div>
`;

exports[`ReactFieldBase does not render the label if showLabel is false 1`] = `
<div>
  <div
    aria-labelledby="undefined_fieldDetails"
    class="form-group hide"
  >
    <input
      name="undefined_edited"
      type="hidden"
      value="false"
    />
    <div
      aria-hidden="true"
      class="form-feedback-group lfr-de__field-feedback"
    >
      <div
        aria-live="assertive"
        class="form-feedback-item"
        id="undefined_fieldError"
      />
    </div>
    <span
      class="sr-only"
      id="undefined_fieldDetails"
    >
      Text
    </span>
  </div>
</div>
`;

exports[`ReactFieldBase renders the FieldBase with contentRenderer 1`] = `
<div>
  <div
    aria-labelledby="undefined_fieldDetails"
    class="form-group hide"
  >
    <div>
      <h1>
        Foo bar
      </h1>
    </div>
    <input
      name="undefined_edited"
      type="hidden"
      value="false"
    />
    <div
      aria-hidden="true"
      class="form-feedback-group lfr-de__field-feedback"
    >
      <div
        aria-live="assertive"
        class="form-feedback-item"
        id="undefined_fieldError"
      />
    </div>
  </div>
</div>
`;

exports[`ReactFieldBase renders the FieldBase with help text 1`] = `
<div>
  <div
    aria-labelledby="undefined_fieldDetails"
    class="form-group hide"
  >
    <input
      name="undefined_edited"
      type="hidden"
      value="false"
    />
    <div
      aria-hidden="true"
      class="form-feedback-group lfr-de__field-feedback"
    >
      <div
        aria-live="assertive"
        class="form-feedback-item"
        id="undefined_fieldError"
      />
      <div
        id="undefined_fieldHelp"
      >
        Type something!
      </div>
    </div>
    <span
      class="sr-only"
      id="undefined_fieldDetails"
    >
      Type something!
    </span>
  </div>
</div>
`;

exports[`ReactFieldBase renders the FieldBase with id 1`] = `
<div>
  <div
    aria-labelledby="Id_fieldDetails"
    class="form-group hide"
  >
    <input
      name="undefined_edited"
      type="hidden"
      value="false"
    />
    <div
      aria-hidden="true"
      class="form-feedback-group lfr-de__field-feedback"
    >
      <div
        aria-live="assertive"
        class="form-feedback-item"
        id="Id_fieldError"
      />
    </div>
  </div>
</div>
`;

exports[`ReactFieldBase renders the FieldBase with label 1`] = `
<div>
  <div
    class="form-group hide"
  >
    <label
      aria-labelledby="undefined_fieldDetails"
      class="ddm-label"
    >
      Text
    </label>
    <input
      name="undefined_edited"
      type="hidden"
      value="false"
    />
    <div
      aria-hidden="true"
      class="form-feedback-group lfr-de__field-feedback"
    >
      <div
        aria-live="assertive"
        class="form-feedback-item"
        id="undefined_fieldError"
      />
    </div>
    <span
      class="sr-only"
      id="undefined_fieldDetails"
    >
      Text
    </span>
  </div>
</div>
`;

exports[`ReactFieldBase renders the FieldBase with required 1`] = `
<div>
  <div
    class="form-group hide"
  >
    <label
      aria-labelledby="undefined_fieldDetails"
      class="ddm-label"
    >
      <span
        class="ddm-label-required reference-mark"
      >
        <svg
          class="lexicon-icon lexicon-icon-asterisk"
          role="presentation"
        >
          <use
            href="#asterisk"
          />
        </svg>
      </span>
    </label>
    <input
      name="undefined_edited"
      type="hidden"
      value="false"
    />
    <div
      aria-hidden="true"
      class="form-feedback-group lfr-de__field-feedback"
    >
      <div
        aria-live="assertive"
        class="form-feedback-item"
        id="undefined_fieldError"
      />
    </div>
    <span
      class="sr-only"
      id="undefined_fieldDetails"
    >
      required
    </span>
  </div>
</div>
`;

exports[`ReactFieldBase renders the add button when repeatable is true 1`] = `
<div>
  <div
    class="form-group hide"
  >
    <div
      class="lfr-ddm-form-field-repeatable-toolbar"
    >
      <button
        aria-label="add-duplicate-field"
        class="ddm-form-field-repeatable-add-button p-0 btn btn-sm btn-primary"
        title="duplicate"
        type="button"
      >
        <svg
          class="lexicon-icon lexicon-icon-plus"
          role="presentation"
        >
          <use
            href="#plus"
          />
        </svg>
      </button>
    </div>
    <label
      aria-labelledby="undefined_fieldDetails"
      class="ddm-empty ddm-repeatable"
    />
    <input
      name="undefined_edited"
      type="hidden"
      value="false"
    />
    <div
      aria-hidden="true"
      class="form-feedback-group lfr-de__field-feedback"
    >
      <div
        aria-live="assertive"
        class="form-feedback-item"
        id="undefined_fieldError"
      />
    </div>
    <span
      class="sr-only"
      id="undefined_fieldDetails"
    >
      Text
    </span>
  </div>
</div>
`;

exports[`ReactFieldBase renders the default markup 1`] = `
<div>
  <div
    aria-labelledby="undefined_fieldDetails"
    class="form-group hide"
  >
    <input
      name="undefined_edited"
      type="hidden"
      value="false"
    />
    <div
      aria-hidden="true"
      class="form-feedback-group lfr-de__field-feedback"
    >
      <div
        aria-live="assertive"
        class="form-feedback-item"
        id="undefined_fieldError"
      />
    </div>
  </div>
</div>
`;
